### DISCUSSION TITLE
Is there a way to keep the generated temporary files by robolectric?

### DISCUSSION THREAD

**Discussion by XuanGao2023** [robolectric/robolectric#7124] [Q&A]
I need to use temporary files to analyze a problem.
After running the test case, I found that the temporary files generated are automatically cleared。
Is there a way to keep the generated temporary files by robolectric?

Thanks

**Comment by utzcoz** [comment]
As I know, Robolectric don't have config to keep temporary files created by internal simulation, and remove it after using to avoid resource leaks(although those temporary files will removed by Linux automatically when they reach recycle state). If you want to analyze content of temporary files, maybe adding proper shadow APIs to exposure those data when running test for later usage, e.g. storing those data to files with your custom code for your need. Could you explain more about your analyzing area/occasion? Maybe we can discuss more about potential shadow APIs or other methods to meet your need instead of keeping temporary files.

**Reply by XuanGao2023** [reply]
For example，after a test case(precisely an operation) runs, save all temporary files and database files of the current simulation process for problem analysis or for later usage. These data may be some abnormally wrong data generated intentionally. The exported data can be used as initialization data for some later test cases, and even the real machine data can be imported into robolectric as initialization data to parse and analyze problems or to do test case detection.

**Reply by utzcoz** [reply]
IMHO, it's difficult for Robolectric. It's recommended to collect those data when running test, and packing it to specific files by your tests, e.g. storing them in map or list, and saving those data at `AfterClass`. Robolectric has massive APIs to get state/data when running tests. If you found it misses some APIs for specific data, it's welcome to send PR to support it. There are many places can generate temporary files in Robolectric, it's not ideal to import them. Sorry for that.

**Comment by hoisie** [comment]
@Pan-Gao there may be some workarounds

1) Add an `@After` block to your tests that copy the temporary files to somewhere else for investigation. You may also be able to use a custom test runner or a custom test application to make this easier for all tests.
2) Build a custom version of Robolectric that does not delete the temporary files and use that.
